lib/paper_trail/request.rb

```diff
@@ -0,0 +1,152 @@
+    # @api public
+    def whodunnit=(value)
+      paper_trail_store[:whodunnit] = value
+    end
+
+    # If nothing passed, returns who is reponsible for any changes that occur.
+    #
+    #   PaperTrail.request.whodunnit = "someone"
+    #   PaperTrail.request.whodunnit # => "someone"
+    #
+    # If value and block passed, set this value as whodunnit for the duration of the block
+    #
+    #   PaperTrail.request.whodunnit("me") do
+    #     puts PaperTrail.request.whodunnit # => "me"
+    #   end
+    #
+    # @api public
+    def whodunnit(value = nil)
+      config_handler(:whodunnit, value) do
+        yield
+      end
+    end
+
+    # Sets any information from the controller that you want PaperTrail to
+    # store.  By default this is set automatically by a before filter.
+    # @api public
+    def controller_info=(value)
+      paper_trail_store[:controller_info] = value
+    end
+
+    # If nothing passed, returns the paper trail info from the controller that you want PaperTrail
+    # to store
+    # See `PaperTrail::Rails::Controller#info_for_paper_trail`.
+    #
+    #   PaperTrail.request.controller_info = { ip: request_user_ip }
+    #   PaperTrail.request.controller_info # => { ip: '127.0.0.1' }
+    #
+    # If value and block passed, set this value as the controller info for the duration of the block
+    #
+    #   PaperTrail.request.controller_info({ ip: '127.0.0.1' }) do
+    #     puts PaperTrail.request.controller_info # => { ip: '127.0.0.1' }
+    #   end
+    #
+    # @api public
+    def controller_info(value = nil)
+      config_handler(:controller_info, value) do
+        yield
+      end
+    end
+
+    # @api public
+    def transaction_id
+      paper_trail_store[:transaction_id]
+    end
+
+    # @api public
+    def transaction_id=(id)
+      paper_trail_store[:transaction_id] = id
+    end
+
+    # TODO: Make this private and have the request method take in a block
+    # Allows for paper trail settings to be set with a block
+    #
+    # config = { whodunnit: 'system', controller_info: { ip: '127.0.0.1' } }
+    # PaperTrail.with_paper_trail_config(config) do
+    #   puts PaperTrail.request.controller_info # => { ip: '127.0.0.1' }
+    #   puts PaperTrail.request.whodunnit # => 'system'
+    # end
+    #
+    # @api public
+    def with_paper_trail_config(config)
+      raise ArgumentError, "no block given" unless block_given?
+
+      previous_config = {}
+      config.each do |config_key, config_value|
+        if config_value
+          previous_config[config_key] = paper_trail_store[config_key]
+          paper_trail_store[config_key] = config_value
+        elsif paper_trail_store[config_key].respond_to?(:call)
+          paper_trail_store[config_key].call
+        end
+      end
+
+      begin
+        yield
+      ensure
+        previous_config.each do |config_key, config_value|
+          paper_trail_store[config_key] = config_value
+        end
+      end
+    end
```

Comment:
Do we need both this *and* `PaperTrail.request({}) do`?

Reply:
I was thinking that `PaperTrail.request({}) do` would just delegate to this method.  Ideally with this new API, people would actually just call the paper trail request class API directly rather than through paper trail, so maybe `PaperTrail.request(...)` doesn't make sense as a public API, and instead they should do `PaperTrail.request.with_config({...})`.  What do you think?